powershell_completions.go

```diff
@@ -11,90 +8,278 @@ import (
+        }
+    }
+
+    if ((($Directive -band $ShellCompDirectiveFilterFileExt) -ne 0 ) -or 
+       (($Directive -band $ShellCompDirectiveFilterDirs) -ne 0 ))  {
+        __%[1]s_debug "ShellCompDirectiveFilterFileExt ShellCompDirectiveFilterDirs are not supported"
+
+        # return here to prevent the completion of the extensions
+        return
+    }
+
+    $Values = $Values | Where-Object {
+        # filter the result
+        $_.Name -like "$WordToComplete*"
 
-	var subCommandCases bytes.Buffer
-	generatePowerShellSubcommandCases(&subCommandCases, c, "")
-	fmt.Fprintf(buf, powerShellCompletionTemplate, c.Name(), c.Name(), subCommandCases.String())
+        # Join the flag back if we have a equal sign flag
+        if ( $IsEqualFlag ) {
+            __%[1]s_debug "Join the equal sign flag back to the completion value"
+            $_.Name = $Flag + "=" + $_.Name
+        }
+    } 
+
+    # Get the current mode
+    $Mode = (Get-PSReadLineKeyHandler | Where-Object {$_.Key -eq "Tab" }).Function
+    __%[1]s_debug "Mode: $Mode"
+
+    $Values | ForEach-Object {
+
+        # store temporay because switch will overwrite $_
+        $comp = $_
+
+        # Powershell supports three different completion modes
+        # - TabCompleteNext (default windows style - on each key press the next option is displayed)
+        # - Complete (works like bash)
+        # - MenuComplete (works like zsh)
+        # You set the mode with Set-PSReadLineKeyHandler -Key Tab -Function <mode>
+
+        # CompletionResult Arguments:
+        # 1) CompletionText text to be used as the auto completion result
+        # 2) ListItemText   text to be displayed in the suggestion list
+        # 3) ResultType     type of completion result
+        # 4) ToolTip        text for the tooltip with details about the object
+
+        switch ($Mode) {
+
+            # bash like
+            "Complete" {
+
+                if ($Values.Length -eq 1) {
+                    __%[1]s_debug "Only one completion left"
 
+                    # insert space after value
+                    [System.Management.Automation.CompletionResult]::new($($comp.Name | __%[1]s_escapeStringWithSpecialChars) + $Space, "$($comp.Name)", 'ParameterValue', "$($comp.Description)")
+
+                } else {
+                    # Add the proper number of spaces to align the descriptions
+                    while($comp.Name.Length -lt $Longest) {
+                        $comp.Name = $comp.Name + " "
+                    }
+
+                    # Check for empty description and only add parentheses if needed
+                    if ($($comp.Description) -eq " " ) {
+                        $Description = ""
+                    } else {
+                        $Description = "  ($($comp.Description))"
+                    }
+
+                    [System.Management.Automation.CompletionResult]::new("$($comp.Name)$Description", "$($comp.Name)$Description", 'ParameterValue', "$($comp.Description)")
```

Comment:
Why set `completionText` param to `"$($comp.Name)$Description"`? 
I couldn't figure out when I checked the code, but this seems to be the only constructor that adds a `$Description` to `completionText` parameter, is there any special meaning to it?

```pwsh
 $myParams = @{
  Native=$true
  CommandName='Foo'
}

$comps=@(
  @{n='add';d=':Add Foo'}
  @{n='list';d=':List Foo'}
  @{n='remove';d=':Remove Foo'}
  @{n='v';d=''}
)

Register-ArgumentCompleter @myParams -ScriptBlock {
  param($WordToComplete,$CommandAst,$CursorPosition)
    $comps.where{$_.n -match "^$WordToComplete"}  | % {
      $Mode = (Get-PSReadLineKeyHandler | Where-Object {$_.Key -eq "Tab" }).Function
        
      [CompletionResult]::new("$($_.n) $($_.d)", "$($_.n) $($_.d)", [CompletionResultType]::ParameterValue, " ")
    }
}
```

```console
~
❯ foo afdaf
add :Add Foo        list :List Foo      remove :Remove Foo  v

~
❯ foo add :Add Foo
```

Reply:
As you might know powershell support 3 different completion modes:
- `TabCompleteNext`
- `Complete`
- `MenuComplete`

By default (powershell) only `MenuComplete` will display a description below the suggestions.
Cobra allows descriptions in autocompletions for zsh and fish. PR #1146 adds description support to bash via a little hack. I adopted this hack here to also allow description to be displayed with the `Complete` mode since powershell does not support this natively . The idea is that the descriptions are appended to the real completion values. However if only one matching completion is left the script wont append the description so that the shell can autocomplete the correct value.

Users can choose if they want descriptions to be displayed while generating the scripts with `GenPowerShellCompletion` for no descriptions and  `GenPowerShellCompletionWithDesc` for descriptions.
```
# With descriptions and Mode 'Complete'
$ helm s[tab]
search  (search for a keyword in charts)  show  (show information of a chart)  status  (displays the status of the named release)
# if you type 'se' you will get this autocompleted to 'search'

# With descriptions and Mode 'MenuComplete' The description of the current selected value will be displayed below the suggestions.
$ helm s[tab]
search    show     status  
search for a keyword in charts

# Without descriptions
$ helm s[tab]
search  show  status
```

Reply:
Nice idea!!
I understand the intent of the code.
This is not a problem.